presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -352,4 +472,316 @@ private static void failWithCorruptionException(boolean failOnCorruptedParquetSt
+        }
+
+        @Override
+        public boolean keep(T value)
+        {
+            if (value == null && !columnDomain.isNullAllowed()) {
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public boolean canDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistic)
+        {
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
+                    Integer min = (Integer) statistic.getMin();
+                    Integer max = (Integer) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics((long) min, (long) max));
+                }
+                else if (statistic.getMin() instanceof Long) {
+                    Long min = (Long) statistic.getMin();
+                    Long max = (Long) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics(min, max));
+                }
+                else if (statistic.getMin() instanceof Float) {
+                    Integer min = floatToRawIntBits((Float) statistic.getMin());
+                    Integer max = floatToRawIntBits((Float) statistic.getMax());
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics((long) min, (long) max));
+                }
+                else if (statistic.getMin() instanceof Double) {
+                    Double min = (Double) statistic.getMin();
+                    Double max = (Double) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetDoubleStatistics(min, max));
+                }
+                else if (statistic.getMin() instanceof Binary) {
+                    Binary min = (Binary) statistic.getMin();
+                    Binary max = (Binary) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetStringStatistics((Slices.wrappedBuffer(min.getBytes())), Slices.wrappedBuffer(max.getBytes())));
+                }
+                //TODO: Add other types
+            }
+            return false;
+        }
+
+        @Override
+        public boolean inverseCanDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistics)
+        {
+            // !canDrop() cannot be used because it might not be correct. To be safe, we just keep the record by returning false.
+            // Since we don't use LogicalNotUserDefined, this method is not called.
+            return false;
+        }
+
+        private boolean canDropCanWithRangeStats(ParquetRangeStatistics parquetStatistics)
+        {
+            Domain domain = createDomain(columnDomain.getType(), true, parquetStatistics);
+            if (columnDomain.intersect(domain).isNone()) {
+                return true;
+            }
+            return false;
+        }
+    }
+
+    class ColumnIndexValueConverter
```

Comment:
shall we consider put `ParquetUserDefinedPredicateTupleDomain`, and `ColumnIndexValueConverter` as standalone class?

Reply:
It is used just inside this file.